### PR TITLE
Adding support for Rotary Position Embeddings

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
[Design doc](https://www.notion.so/Adding-support-for-Rotary-Position-Embeddings-6b17c27c26f94233bed5b63c010c6de2)

Experiments: [125M model](https://wandb.ai/mosaic-ml/rope_v_alibi_eval_125m?workspace=user-shashank-rajput), [1B model](https://wandb.ai/mosaic-ml/rope_v_alibi_eval_1b?workspace=user-shashank-rajput).

### CODE REVIEW

File: llmfoundry/models/layers/attention.py [hunk 1]
```diff
@@ -539,6 +541,24 @@ def forward(
...
+            seq_len = rotary_emb_w_offset_info['seq_len']
+            pos = rotary_emb_w_offset_info['pos']
+            (cos, sin) = rotary_emb(query, seq_len)
+            query, key = apply_rotary_pos_emb(query, key, cos, sin, pos)
+
+            query = query.transpose(1, 2)
+            key = key.transpose(1, 2)
+            query = query.reshape(*(query.shape[:-2]), self.d_model)
+            key = key.reshape(*(key.shape[:-2]),
+                              self.kv_n_heads * self.head_dim)
```
Comment (vchiley) [inline llmfoundry/models/layers/attention.py:560]: If this was moved into `scaled_multihead_dot_product_attention`, `flash_attn_fn`, and `triton_flash_attn_fn`, could we then avoid some of the transpose (then un-transpose)?

File: llmfoundry/models/layers/attention.py [hunk 2]
```diff
@@ -5,13 +5,14 @@
 
 import math
 import warnings
-from typing import Any, List, Optional, Tuple
+from typing import Any, Dict, List, Optional, Tuple
```
Comment (vchiley) [inline llmfoundry/models/layers/attention.py:8]: we can probably just use `dict` as the type instead of needing to import `Dict`
same thing can probably be said for `List` and `Tuple` (but if you want to say thats outside the scope of this PR thats fine)

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -397,12 +428,16 @@ def forward(
...
                     min=0,
                 )
+            if self.rope:
+                rotary_emb_w_offset_info = {
+                    'rotary_emb': self.rotary_embedding,
+                    'pos': pos,
+                    'seq_len': S + past_position
+                }
+            if self.learned_pos_emb:
+                pos_emb = self.wpe(pos)
```
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:438]:
```suggestion
            x = tok_emb
            if self.learned_pos_emb:
                pos_emb = self.wpe(pos)
                x = x + pos_emb
```

and remove 
```
pos_emb = 0.0
```
above 
and remove 
```
x = tok_emb + pos_emb
```
below

File: scripts/inference/run_mpt_with_ft.py [hunk 4]
```diff
@@ -280,7 +280,7 @@ def main():
     shared_contexts_ratio = args.shared_contexts_ratio
     layernorm_eps = args.layernorm_eps
     use_attention_linear_bias = args.alibi
-    has_positional_encoding = not args.alibi
+    has_positional_encoding = not args.alibi  # TODO: Should probably be: has_positional_encoding = not (args.alibi or args.rope)
```
Comment (vchiley) [inline scripts/inference/run_mpt_with_ft.py:283]: are we leaving this as a todo? or asking @dskhudia to comment on what should be done here?

File: llmfoundry/models/layers/attention.py (hunk 2 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/layers/attention.py]: Fixed! (including the List and Tuple)

File: scripts/inference/run_mpt_with_ft.py (hunk 4 above)
Comment (ShashankMosaicML) [reply scripts/inference/run_mpt_with_ft.py:283]: I checked, this should actually be `not args.alibi`. This is because this argument is (only) used for instantiating [ParallelGPT](https://github.com/NVIDIA/FasterTransformer/blob/df4a7534860137e060e18d2ebf019906120ea204/examples/pytorch/gpt/utils/gpt.py#L441), which seems to only support learned positional encodings or attention linear biases (alibi).

Comment (vchiley) [reply scripts/inference/run_mpt_with_ft.py:283]: So the added `TODO` should be removed?

Comment (ShashankMosaicML) [reply scripts/inference/run_mpt_with_ft.py]: Yes, removed it in the latest commit.

File: llmfoundry/models/layers/attention.py [hunk 5]
```diff
@@ -34,6 +33,13 @@ def is_flash_v1_installed():
     return version.parse(flash_attn.__version__) < version.parse('2.0.0')
 
 
+if is_flash_v1_installed():
+    import transformers
+    transformers.utils.is_flash_attn_available = lambda: False
```
Comment (vchiley) [inline llmfoundry/models/layers/attention.py:38]: @dakinggg dont you have a util for checking if flash v1 vs flashv2 is installed?

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 6]
```diff
@@ -247,8 +247,8 @@ def _validate_config(self) -> None:
                 self.attn_config['rope_imp']
                 == 'dail') and (self.attn_config['rope_dail_config']['type']
                                 not in ['original', 'xpos']):
-            raise NotImplementedError(
-                'If using dail implementation of rope, the type should be one of "original" or "xpos".'
+            raise ValueError(
+                'If using the dail implementation of rope, the type should be one of "original" or "xpos".'
```
Comment (vchiley) [inline llmfoundry/models/mpt/configuration_mpt.py:251]: add using RoPE v ALiBi section in [tutorial](https://github.com/mosaicml/llm-foundry/blob/main/TUTORIAL.md#faqs)
Document the different configurations there.

**Review by vchiley — COMMENTED** [review]
Documentation request added

File: llmfoundry/models/layers/blocks.py [hunk 7]
```diff
@@ -41,6 +41,18 @@ def __init__(
...
+                'rope_imp': 'dail',
+                'rope_dail_config': {
+                    'type': 'original',
+                    'pos_idx_in_fp32': True,
+                    'xpos_scale_base': 512,
+                },
+                'rope_hf_config': {
+                    'type': 'no_scaling',
+                    'factor': 1.0,
+                },
```
Comment (vchiley) [inline llmfoundry/models/layers/blocks.py:55]:
```suggestion
```

File: tests/test_flash_triton_torch.py [hunk 8]
```diff
@@ -18,26 +21,66 @@ def allclose_helper(t0: torch.Tensor,
...
+}, {
+    'alibi': False,
+    'rope': True,
+    'rope_theta': 10000,
+    'rope_imp': 'hf',
+    'rope_dail_config': {
+        'type': 'original',
+        'pos_idx_in_fp32': True,
+        'xpos_scale_base': 512,
+    },
```
Comment (vchiley) [inline tests/test_flash_triton_torch.py:53]:
```suggestion
```

File: tests/test_flash_triton_torch.py [hunk 9]
```diff
@@ -18,26 +21,66 @@ def allclose_helper(t0: torch.Tensor,
...
+    'rope_imp': 'dail',
+    'rope_dail_config': {
+        'type': 'original',
+        'pos_idx_in_fp32': True,
+        'xpos_scale_base': 512,
+    },
+    'rope_hf_config': {
+        'type': 'no_scaling',
+        'factor': 1.0,
+    },
```
Comment (vchiley) [inline tests/test_flash_triton_torch.py:43]:
```suggestion
```

**Review by vchiley — COMMENTED** [review]
rm few extra config setting to avoid confusion

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 6 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py:254]: Added [here](https://github.com/ShashankMosaicML/llm-foundry/blob/rotary_hf_imp/TUTORIAL.md#what-kinds-of-positional-embeddings-does-llm-foundry-support)

File: tests/test_flash_triton_torch.py (hunk 9 above)
Comment (ShashankMosaicML) [reply tests/test_flash_triton_torch.py]: Done

File: tests/test_flash_triton_torch.py (hunk 8 above)
Comment (ShashankMosaicML) [reply tests/test_flash_triton_torch.py]: Done

**Comment by ShashankMosaicML** [comment]
@dakinggg will this merge cause any problems in merging mpt code into the Hugging Face github codebase?

**Review by vchiley — APPROVED** [review]

File: llmfoundry/models/layers/attention.py [hunk 10]
```diff
@@ -33,6 +33,13 @@ def is_flash_v1_installed():
     return version.parse(flash_attn.__version__) < version.parse('2.0.0')
 
 
+if is_flash_v1_installed():
```
Comment (dakinggg) [inline llmfoundry/models/layers/attention.py:36]: leave a comment behind about why this is necessary

File: llmfoundry/models/layers/attention.py [hunk 11]
```diff
@@ -581,6 +589,40 @@ def forward(
             query = self.q_ln(query).to(dtype)
             key = self.k_ln(key).to(dtype)
 
+        if rotary_emb_w_meta_info is not None:
+            rotary_emb = rotary_emb_w_meta_info['rotary_emb']
+            seq_len = rotary_emb_w_meta_info['seq_len']
+            offset_info = rotary_emb_w_meta_info['offset_info']
+
+            query = query.view(*(query.shape[:-1]), -1, self.head_dim)
```
Comment (dakinggg) [inline llmfoundry/models/layers/attention.py:597]: instead of these * unpacking for shapes, can you define variables for each dimension you need and use them directly?

File: llmfoundry/models/layers/blocks.py [hunk 12]
```diff
@@ -41,6 +41,18 @@ def __init__(
                 'attn_uses_sequence_id': False,
                 'alibi': False,
                 'alibi_bias_max': 8,
+                'rope': False,
+                'rope_theta': 10000,
+                'rope_imp': 'dail',
+                'rope_dail_config': {
```
Comment (dakinggg) [inline llmfoundry/models/layers/blocks.py:47]: do we need two separate rope configs? or can we just have `rope_config`?

File: llmfoundry/models/layers/blocks.py [hunk 13]
```diff
@@ -41,6 +41,18 @@ def __init__(
                 'attn_uses_sequence_id': False,
                 'alibi': False,
                 'alibi_bias_max': 8,
+                'rope': False,
```
Comment (dakinggg) [inline llmfoundry/models/layers/blocks.py:44]: is there any reason to not import this default attn config from configuration_mpt?

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 14]
```diff
@@ -19,6 +19,18 @@
     'attn_uses_sequence_id': False,
     'alibi': False,
     'alibi_bias_max': 8,
+    'rope': False,
+    'rope_theta': 10000,
+    'rope_imp': 'dail',
+    'rope_dail_config': {
```
Comment (dakinggg) [inline llmfoundry/models/mpt/configuration_mpt.py:25]: same comment on configs

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 15]
```diff
@@ -94,6 +106,16 @@ def __init__(
                     Defaults to ``False`` meaning any provided `sequence_id` will be ignored.
                 alibi (bool): Whether to use the alibi bias instead of position embeddings.
                 alibi_bias_max (int): The maximum value of the alibi bias.
+                rope (bool): Whether to use rotary positional embeddings.
+                rope_theta (int): The base frequency for rope.
+                rope_imp (str): The implementation of rope to use. One of 'hf' (to use the implementation from https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py) or 'dail' (to use the implementation from https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/layers/rotary.py).
```
Comment (dakinggg) [inline llmfoundry/models/mpt/configuration_mpt.py:111]: lets rename this `rope_impl`

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 16]
```diff
@@ -150,10 +172,12 @@ def __init__(
             del kwargs['name']
         if 'loss_fn' in kwargs:
             del kwargs['loss_fn']
-        if self.attn_config.get('alibi', False):
+        if self.attn_config.get('alibi', False) or self.attn_config.get(
```
Comment (dakinggg) [inline llmfoundry/models/mpt/configuration_mpt.py:175]: lets also check that alibi and rope are not both set?

File: llmfoundry/models/mpt/configuration_mpt.py [hunk 17]
```diff
@@ -164,6 +188,9 @@ def _set_config_defaults(self, config: Dict[str, Any],
         for k, v in config_defaults.items():
             if k not in config:
                 config[k] = v
+            elif isinstance(v, dict):
```
Comment (dakinggg) [inline llmfoundry/models/mpt/configuration_mpt.py:191]: why was this recursion necessary?

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 18]
```diff
@@ -70,6 +86,50 @@
 log = logging.getLogger(__name__)
 
 
+def _rotary_embedding(config: MPTConfig):
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:89]: please don't pass the full config in. just explicitly pass in the pieces you need

File: tests/test_flash_triton_torch.py [hunk 19]
```diff
@@ -18,26 +21,57 @@ def allclose_helper(t0: torch.Tensor,
...
+    rope.
     """
     from llmfoundry.models.layers import attention
-
+    alibi = pos_emb_config['alibi']
+    rope = pos_emb_config['rope']
     if alibi and (attn_impl_0 == 'flash' or attn_impl_1 == 'flash'):
         pytest.xfail('flash attn does not support alibi')
 
+    if rope and (pos_emb_config['rope_imp']
```
Comment (dakinggg) [inline tests/test_flash_triton_torch.py:71]: can we make this case raise a nice error that tells you want to do to fix it?

File: tests/test_rope_dail_vs_hf.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+from omegaconf import OmegaConf as om
+from transformers.models.llama.modeling_llama import \
+    LlamaDynamicNTKScalingRotaryEmbedding as HFDynamicNTKScalingRotaryEmbedding
+from transformers.models.llama.modeling_llama import \
+    LlamaLinearScalingRotaryEmbedding as HFLinearScalingRotaryEmbedding
+from transformers.models.llama.modeling_llama import \
+    LlamaRotaryEmbedding as HFRotaryEmbedding
+
+
+def gen_rotary_embedding(rope_head_dim: int, pos_emb_config: dict,
```
Comment (dakinggg) [inline tests/test_rope_dail_vs_hf.py:21]: this function looks like it can be imported from the one you use in the actual code. if so, lets do that

File: tests/test_model.py [hunk 21]
```diff
@@ -612,23 +646,42 @@ def test_forward_with_padding(attention_impl: str, device: str, alibi: bool):
                              attention_mask=batched_attention_mask).logits
 
         # check that right padding and left padding produce the same output
-        assert torch.allclose(right_padding_output[0, :3],
-                              left_padding_output[0, 3:],
-                              atol=1e-6 if attention_impl == 'torch' else 1e-8)
-        if not alibi:
+        if rope and pos_emb_config[
```
Comment (dakinggg) [inline tests/test_model.py:649]: lets just define the expected tolerance in the if statement instead of the full assert

File: llmfoundry/models/layers/attention.py (hunk 10 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/layers/attention.py:36]: Done.

File: tests/test_model.py (hunk 21 above)
Comment (ShashankMosaicML) [reply tests/test_model.py]: Done.

File: tests/test_flash_triton_torch.py (hunk 19 above)
Comment (ShashankMosaicML) [reply tests/test_flash_triton_torch.py]: Done, [here](https://github.com/mosaicml/llm-foundry/blob/ac0fd40651a504ac4a902464e4009fdf9698402b/llmfoundry/models/mpt/configuration_mpt.py#L244)

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 18 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/modeling_mpt.py]: Done, thanks for the suggestion!

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 15 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py]: Done!

File: llmfoundry/models/layers/blocks.py (hunk 13 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/layers/blocks.py]: Done, thanks for the suggestion!

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 16 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py:161]: Alibi and rope are not mutually exclusive, someone can turn on both alibi and rope.

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 14 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py]: Please see my response [above](https://github.com/mosaicml/llm-foundry/pull/675#discussion_r1381979636).

File: tests/test_rope_dail_vs_hf.py (hunk 20 above)
Comment (ShashankMosaicML) [reply tests/test_rope_dail_vs_hf.py]: I did not originally import it because I wanted this test to be as independent of the actual code as possible (since this only checks if dail and hf implementations produce the same results). However, importing the function will also test the function indirectly, so I will import it now. 
**tldr**: Done

**Review by dakinggg — APPROVED** [review]

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 22]
```diff
@@ -378,31 +451,44 @@ def forward(
...
+                    # adjust the position indices to account for padding tokens
+                    pos = torch.clamp(
+                        pos - torch.cumsum((~attention_mask).to(torch.int32),
+                                           dim=1)[:, past_position:],
+                        min=0,
+                    )
+                if self.learned_pos_emb:
+                    x = x + self.wpe(pos)
+                elif self.rope and self.rope_impl == 'hf':
+                    rotary_emb_w_meta_info = {
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/modeling_mpt.py:479]: Done, thanks for pointing out!

File: llmfoundry/models/layers/attention.py [hunk 23]
```diff
@@ -581,6 +593,41 @@ def forward(
...
+            rotary_emb = rotary_emb_w_meta_info['rotary_emb']
+            seq_len = rotary_emb_w_meta_info['seq_len']
+            offset_info = rotary_emb_w_meta_info['offset_info']
+            assert query.shape[:2] == key.shape[:2]
+            assert query.shape[:2] == value.shape[:2]
+            bsz, seqlen = query.shape[:2]
+            query = query.view(bsz, seqlen, -1, self.head_dim)
+            key = key.view(bsz, seqlen, -1, self.head_dim)
+
+            if rotary_emb_w_meta_info['imp'] == 'dail':
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply llmfoundry/models/layers/attention.py]: Done, thanks for pointing out!

File: llmfoundry/models/layers/attention.py [hunk 24]
```diff
@@ -581,6 +593,41 @@ def forward(
             query = self.q_ln(query).to(dtype)
             key = self.k_ln(key).to(dtype)
 
+        if rotary_emb_w_meta_info is not None:
+            rotary_emb = rotary_emb_w_meta_info['rotary_emb']
+            seq_len = rotary_emb_w_meta_info['seq_len']
+            offset_info = rotary_emb_w_meta_info['offset_info']
+            assert query.shape[:2] == key.shape[:2]
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply llmfoundry/models/layers/attention.py]: Done, it was just a sanity check.

File: llmfoundry/models/mpt/configuration_mpt.py (hunk 16 above)
Comment (ShashankMosaicML) [reply llmfoundry/models/mpt/configuration_mpt.py:161]: Sorry, missed this comment. No, I don't think anybody has done this. I tried it [here](https://wandb.ai/mosaic-ml/rope_stability_lr_bs_125m/reports/loss-train-total-23-11-04-19-03-41---Vmlldzo1ODc0MDA5) and it gave slightly faster convergence initially, but converged to the same final training loss:

image
